### ISSUE TITLE
Memory.Initialization: needsDeinit and explicitDeinit on unmanaged classes?

### ISSUE THREAD

**Issue by benharsh** [chapel-lang/chapel#20410]
Currently the ``needsDeinit`` function in the Memory.Initialization module returns ``false`` when given an unmanaged class type. The ``explicitDeinit`` function will not attempt to delete an unmanaged class. Here's a sample program demonstrating this behavior:

```chpl
use Memory.Initialization;

class C {
  var x : int = 5;

  proc deinit() {
    writeln("goodbye");
  }
}

proc main() {
  var c = new unmanaged C();
  writeln(needsDeinit(c.type)); // false
  writeln("-----");
  explicitDeinit(c); // nothing happens
  writeln("-----");
  delete c; // "goodbye"
}
```

The program above results in the following output:
```
false
-----
-----
goodbye
```

So far ``needsDeinit`` returns true if something would be deinitialized automatically by the compiler (i.e. records and *managed* classes).

**Should this behavior change?**

At the very least we should document this current behavior as it might surprise some users.

**Comment by benharsh** [comment]
Personally I don't feel that a change is needed here provided we document this behavior, but I wanted to start a discussion because I don't think this has been pointed out yet.

**Comment by bradcray** [comment]
I wasn't completely surprised that `needsDeinit()` would return `false` for unmanaged classes, though I think `true` probably makes more sense.  Though I can also understand why it's as it is.  Something like `compilerTypicallyDeinitializes()` feels less ambiguous and more like what we've got today (though that's obviously a mouthful).

I was much more surprised that `explicitDeinit()` would not call `deinit()` on the object.

**Comment by mppf** [comment]
I think the current behavior makes sense. My view is that unmanaged classes don't "need to be deinited" but they "need to be deleted".

But, the fact that classes can have a `proc deinit` makes this a bit weirder. Consider even the case of an `owned C` where we have a `class C`. There are actually *two* deinits we could be talking about:
 * the `owned` deinit
 * the `C` deinit

I'm not sure if we need to change these function names but in Memory.Initialization, I think it should only talk about the first of those two. Given that, my view is that there is no relevant "deinit" needed for `unmanaged` classes.

Another case to consider (for terminology purposes at least) is what these functions do for borrowed classes. Like an unmanaged class, a borrowed class can have a `deinit` method. But I think that is irrelevant for the goals of needsDeinit and explicitDeinit and that borrowed classes should be treated the same as an `int` (i.e. needsDeinit returns false and explicitDeinit does nothing).

**Comment by bradcray** [comment]
Based on Michael's comment, would it make sense to use a term other than 'deinit' in these routines given that we're not talking about the user-facing routine named `deinit` so much as a property of the type itself?  E.g., a term to impart "needs scope-driven destruction / deallocation / cleanup" which I think is what we're talking about given the clarification just above?

**Comment by mppf** [comment]
In the spec, we call it "deinitialization" which occurs at the "deinitialization point" / "deinit point".

If we want to better distinguish between these two, I think it would be reasonable to not shorten the name of these to "deinit" but that too is potentially confusing.

Another strategy is to use two different terms - e.g. We could use "destruction" to describe the calls to `proc deinit` and "deinitialization" to describe what a `proc deinit` does.

**Comment by benharsh** [comment]
Maybe something like ``isImplicitlyDestroyed`` and ``explicitDestroy``? In the documentation for ``explicitDestroy`` we could say that it only operates on things that are implicitly destroyed. ``isImplicitlyDestroyed`` could describe the relevant types, or perhaps refer to some language documentation.

**Comment by benharsh** [comment]
Also, I'm imagining that we'll issue a compiler-error for something that we don't expect to implicitly destroy. I think a user would want to know if they tried to call ``explicitDestroy`` on an ``int``, which may indicate they are being a bit sloppy with their memory management (at least in terms of knowing what types they want to support).

Alternatively we could keep silently doing nothing when given a 'wrong' type.

**Comment by bradcray** [comment]
>  I'm imagining that we'll issue a compiler-error for something that we don't expect to implicitly destroy.

I like this idea, mostly because I don't like silent no-op surprises.  Also because, if I'm wrong, it's easier to go from a compilerError to a no-op without breaking code than vice-versa.

I feel the need to wrap that up with my repeated "but I've never used these features, so am not at all expert in them" caveat.

**Comment by mppf** [comment]
> Also, I'm imagining that we'll issue a compiler-error for something that we don't expect to implicitly destroy. I think a user would want to know if they tried to call `explicitDestroy` on an `int`, which may indicate they are being a bit sloppy with their memory management (at least in terms of knowing what types they want to support).
> 
> Alternatively we could keep silently doing nothing when given a 'wrong' type.

IMO the place where I have used these features (or similar undocumented features) is in generic code. In a generic function I might want to "destroy the element if necessary". If `explicitDeinit` or whatever we rename it to just does nothing for types that don't need destruction, then it is easier to write. Vs if it is a compilation error, the generic code always has to use a pattern like `if needsDeinit(elt.type) then explicitDeinit(elt)` which seems a bit wordy.

I don't have a strong opinion here though because these are pretty low level features & it's OK if it is not as short or as easy as possible.

**Comment by bradcray** [comment]
> f it is a compilation error, the generic code always has to use a pattern like if needsDeinit(elt.type) then explicitDeinit(elt) which seems a bit wordy.

I assumed that this was the whole motivation for `needsDeinit()` existing — to guard calls to `explicitDeinit()`.  Are there other motivating use cases for it?  (apart from maybe just debug printing or the like).

**Comment by mppf** [comment]
> > f it is a compilation error, the generic code always has to use a pattern like if needsDeinit(elt.type) then explicitDeinit(elt) which seems a bit wordy.
> 
> I assumed that this was the whole motivation for `needsDeinit()` existing — to guard calls to `explicitDeinit()`. Are there other motivating use cases for it? (apart from maybe just debug printing or the like).

Yes, for example code working with deiniting an array might wish to know if the element type needs deinit before trying to deinit all of the elements (because if the element type does not deinit, there is no need to loop over the elements).

**Comment by benharsh** [comment]
A few things to try and keep this topic moving:

1) I think we're in agreement that these functions shouldn't delete unmanaged or borrowed classes

2) The term "deinit" is a little confusing.
    1) Is there support for ``isImplicitlyDestroyed`` and ``explicitDestroy``? The former would describe the acceptable types. 
    2) An alternative could be ``hasImplicitDestroy``

3) I don't think we've settled on whether ``explicitDeinit`` should issue a compile-time error. I would prefer to go with such an error to avoid silent memory bugs. The trade-off here is extra effort by the author that uses ``explicitDeinit`` in the case of generic types. There's also the possibility that their users might run into errors if they pass in a type that wasn't tested by the author. I think it's worth it to avoid weird, silent memory bugs.

**Comment by bradcray** [comment]
> Is there support for isImplicitlyDestroyed and explicitDestroy? The former would describe the acceptable types.

I could get behind these and definitely prefer the `is` form over `has`.  The main pause I have the names is that it sounds like we're asking "Is this thing implicitly/automatically destroyed?  Yes?  OK, then I'm going to explicitly destroy it!" which sounds... contradictory.  Whereas the `needs...` form told more of a story like "This needs something?  Then I'd better take care of it."

The following has the same problem, but I could imagine `isAutomaticallyDestroyed` or `isAutoDestroyed` as synonyms for the first call (intentionally closing my eyes and ignoring the terminology overlap with our old-school autoDestroy hooks).  I think they're arguably slightly more intuitive names, though the proposal has the benefit of creating symmetry between 'implicit' and 'explicit'.

> I don't think we've settled on whether explicitDeinit should issue a compile-time error. I would prefer to go with such an error to avoid silent memory bugs. The trade-off here is extra effort by the author that uses explicitDeinit in the case of generic types. There's also the possibility that their users might run into errors if they pass in a type that wasn't tested by the author. I think it's worth it to avoid weird, silent memory bugs.

I still tend to lean this way as well, despite Michael's example, primarily because it seems more cautious/explicit, while also allowing us to relax it to the current behavior if it becomes a huge thorn in our side over time.

**Comment by benharsh** [comment]
> though the proposal has the benefit of creating symmetry between 'implicit' and 'explicit'.

What about ``isAutoDestroyed`` and ``manualDestroy``?

**Comment by bradcray** [comment]
I'd be OK with either that or your previous proposal.  They both have the noted downside of seeming to ask a question and then doing something that might seem counter to the answer (i.e., "why are we destroying it if said it's going to be autodestroyed?"), but I don't have any solutions to that.

Mostly, I feel like we're overdue to get a read from @mppf on these latest proposals.

**Comment by mppf** [comment]
It'd be OK with me for `explicitDeinit` (whetever we end up calling it) to raise an error for types that don't need to be explicitly deinited. But, I don't understand why it matters for bugs:

> * I don't think we've settled on whether `explicitDeinit` should issue a compile-time error. I would prefer to go with such an error to avoid silent memory bugs. The trade-off here is extra effort by the author that uses `explicitDeinit` in the case of generic types. There's also the possibility that their users might run into errors if they pass in a type that wasn't tested by the author. I think it's worth it to avoid weird, silent memory bugs.

What kind of weird silent memory bug are you imagining? If the `explicitDeinit` is run on something that is POD or an integer, there is nothing to do, and I'm not immediately seeing the benefit of forcing the programmer to remove the `explicitDeinit` call in that case vs. just having it do nothing. Is your concern that the programmer might be confused and accidentally call `explicitDeinit` on something other than what they meant to?


> I still tend to lean this way as well, despite Michael's example, primarily because it seems more cautious/explicit, while also allowing us to relax it to the current behavior if it becomes a huge thorn in our side over time.

I'm happy with the argument that we could relax it in the future if needed, as a way to make that decision.

> > Is there support for `isImplicitlyDestroyed` and `explicitDestroy`? The former would describe the acceptable types.

> > What about `isAutoDestroyed` and `manualDestroy`?

> I'd be OK with either that or your previous proposal. They both have the noted downside of seeming to ask a question and then doing something that might seem counter to the answer (i.e., "why are we destroying it if said it's going to be autodestroyed?"), but I don't have any solutions to that.

Well, one strategy would be: instead of saying "does it need to be auto destroyed", say "does it do anything when it is deinited".

That would lead to something like `hasEmptyDeinit` / `hasNontrivialDeinit` and `manualDestroy`. We could also/alternatively tie it to "POD"-ness, but I think we probably have been abusing that term already, so it is probably better to have something more specific.

**Comment by bradcray** [comment]
`hasNontrivialDeinit` is intriguing to me, but...  wouldn't this potentially return `true` for a class that had a user-defined `deinit()` procedure?  Am I just back to the two definitions of de-init again?  Yet somehow `hasNontrivialDestroy` or `requiresNontrivialDestroy` doesn't seem quite right either...?

**Comment by mppf** [comment]
it could be something like `hasRecordDeinit` but that might just confuse matters further...

**Comment by bradcray** [comment]
At first blush, that seems pretty attractive to me...  I still feel pretty unfamiliar firsthand with realistic use cases, though, so will be curious what Michael thinks.

In this context, I _think_ I'd even be OK with changing "destroy" back to "deinit" (if desired) because it's less about whether or not the type has a deinit or variables of that type need to be deinitialized, and more whether the requirement is due to the move (I think?).  That said, I'm fine with "destroy" as well.

**Comment by mppf** [comment]
@benharsh - can you show a code example with those function names that would make sense to you? I am not seeing it, at the moment.

**Comment by benharsh** [comment]
```chpl
// Some imaginary record method that deletes an element
// Has an 'eltType' field
// Has a 'table' array that stores a record with the data and full/empty state
proc R.delete(i:int) {
  if movedTypeNeedsDestroy(eltType) then
    destroyMoved(table[i].data);
  table[i].full = false;
}
```

Does that help?

**Comment by mppf** [comment]
I guess the idea of "moved" here is that you would only need to use these functions if you had "moved" the value into a buffer? That seems OK but a bit off-to-the-side of how I would think of these methods.

**Comment by benharsh** [comment]
In my notes I had it down that this was 'complete', but looking over the topic again it's not quite definitive.

Let's see if we can vote on some items here:
👍 - "move" and "destroy": ``movedTypeNeedsDestroy``, ``destroyMoved``
🚀 - "move" and "deinit": ``movedTypeNeedsDeinit``, ``deinitMoved``
👎 - if you think that we shouldn't tie these functions to the concept "moving"

**Comment by mppf** [comment]
> if you think that we shouldn't tie these functions to the concept "moving"

I think this is more or less where I am, but I don't have a better alternative. I am wondering if it would make sense to introduce a term for types where variables of that type are automatically de-initialized when they go out of scope. I would just say it is "record" but we already have the strategy of trying not to consider array, domain, and string as records (rather they happen to be implemented that way) and these types are automatically de-initialized when they go out of scope.

This line of thinking is sortof along the lines of

https://github.com/chapel-lang/chapel/issues/20410#issuecomment-1209668581

> Another strategy is to use two different terms - e.g. We could use "destruction" to describe the calls to `proc deinit` and "deinitialization" to describe what a `proc deinit` does.

It might be good to see if there are better terms. (IIRC Rust uses "drop" for one of these?) But if we went with these terms, I would expect perhaps `needsDestruction` / `needsDestroy` and `destroy` as the function names.

**Comment by benharsh** [comment]
I wanted to check back in here and try to revive the discussion. Have anyone's opinions changed since last looking at the issue?

**Comment by benharsh** [comment]
It was pointed out to me that a summary of sorts might be useful:

Generally we seem to be on the same page that these functions should concern types that are, or should be, destroyed when going out of scope.

There also seems to be agreement that these functions should return false, or do nothing, when the type is unmanaged or borrowed. I'm not sure if we've reached consensus on whether we should issue a compiler error when given a type that would not or cannot destroy (e.g. int). We might reach a decision if we agree that by issuing a compiler error now, we could relax that behavior at a later date.

This led to a discussion about how we probably don't want to use the word "deinit" in the function names since unmanaged or borrowed types might have a user-defined ``deinit``.

I think there are a couple of concepts we've been trying to represent in the function names:
A) Word(s) representing the implicit/out-of-scope behavior
B) Word(s) representing the deinitialization of memory (e.g. "destroy", "destruction")
C) Whether we should tie these functions to the notion of "moving" memory

It has also been proposed that we could try to develop or borrow a term to describe the relevant behavior, which we could then point to in these functions' documentation.

A couple of brief descriptions of example usage:
1) checking whether an array's elements need to be destroyed one-by-one. If not we can avoid a loop over every element.
2) Deleting an entry in some kind of irregular array/table.

**Comment by benharsh** [comment]
After some offline discussion, I believe we're converging on one of the following options:

**Comment by bradcray** [comment]
Capturing some things mentioned to @benharsh last week:

My instinct, given the proposals above, was to abbreviate Deinitialization as Deinit (because it seemed unambiguous and would save a lot of typing).  He explained that the long-form spelling was intentional, to reduce confusion with `deinit` (e.g., does `needsDeinit()` indicate "You should write ` `proc deinit()` for this type"?).  I'm not sure I agree that spelling out `Deinitialization` will reduce that potential confusion since I think of `deinit` as standing for `deinitialize` or `deinitializer`, so I think I'd still equate the two.  GIven these two choices, that makes me lean more toward the "destroy" approach, since it's shorter and less ambiguous w.r.t. `deinit()`.

**Comment by mppf** [comment]
@bradcray - currently, [the spec](https://chapel-lang.org/docs/language/spec/variables.html#variable-lifetimes) uses both "destroy" and "deinitialize" to describe what happens to clean up a specific variable, which we know might end up calling a `proc deinit` in the process. I think we should have consistent terminology here; so if we go with `needsDestroy` / `destroy` (which is totally fine with me) then I think we should update that spec section to use "destroy" instead of "deinitialize".

**Comment by dlongnecke-cray** [comment]
I would prefer the names `isDeinitNontrivial(type t)` and `invokeDeinit(v)`.

I don't think we should avoid having `deinit` in the name here. Functionally all that the `explicitDeinit` function does is call the deinitializer on a value, and it will do that every time you call it.

Avoiding the word deinit and trying to coin a new term might be a bit overkill here as the function isn't really doing something beyond calling deinit, and I don't think we plan for it to. It seems like too much abstraction.

I just think we need to have the function's name be even more explicit about suggesting that it is calling deinit. I think the name `invokeDeinit` does that.

For `isDeinitNontrivial(type t)`, I think the name more strongly associates with the type. One of the issues with `needsDeinit` is that there is a possible conflation with the state of a value (yes, even though the function takes a type). I think that the stronger language in `isDeinitNontrivial` suggests "yes, the deinit for `t` does something and is important to call".

**Comment by benharsh** [comment]
> I don't think we should avoid having deinit in the name here. Functionally all that the explicitDeinit function does is call the deinitializer on a value, and it will do that every time you call it.

Are you of the opinion that these functions should accept and deinitialize an unmanaged class?

**Comment by dlongnecke-cray** [comment]
No, I do not think these functions should deinitialize an `unmanaged`. I think of `delete` as separate from `deinit`. If a user makes a collection of `unmanaged` classes, they would expect the elements to not be deleted by the collection when the collection goes out of scope. If we did so then we would just be making more work by going against the most common use-case.

**Comment by bradcray** [comment]
@mppf / @dlongnecke-cray : Could one of you recap the rationale for why you thought `deinit` was not a good name/term in these routines (which led to `deinitialization`)?  I bought it when I heard it but can't recall it anymore.  (if it's in this thread and I'm forgetting, feel free to point back to it).

**Comment by mppf** [comment]
@dlongnecke-cray - consider an variable of type `shared MyClass`. While you and I know that `shared` is a record and that it contains a `deinit` that will `delete` an instance if the reference count is 0 after decrementing, that's not part of the user view we are trying to present. We are trying to present the view that `shared` is a kind of class type. In that event, it is confusing to say that variable of type `shared` will `deinit` when it goes out of scope. The method `MyClass.deinit` won't be called if the reference count is not zero. So, no user-facing `deinit` will be called at all.

We have the same problem when we describe how arrays and strings work. Again, they are implemented as a record, but we want to view that fact as an implementation detail.

FWIW, I'm pretty far in the camp of "using `deinit` in the name is confusing" on this point.

Responding to @bradcray - the previous conversation is here: https://github.com/chapel-lang/chapel/issues/20410#issuecomment-1208780702 . Besides points along the lines of what I just made, we are concerned that an `unmanaged MyClass` or a `borrowed MyClass` will have the replacement for `needsDeinit` return `false`, but these types do have a related `proc MyClass.deinit()`.

**Comment by dlongnecke-cray** [comment]
I'd prefer `canDeinitialize` and `deinitialize`, then. I like the sound of `canDeinitialize` over `needsDeinitialization` because the former is asking "can I call deinitialize here, for this type?".

**Comment by benharsh** [comment]
Pinging @vasslitvinov , @DanilaFe , and @jeremiah-corrado who expressed interest in this topic offline. No pressure to jump in right away, but if you have any thoughts while the topic is fresh I thought it was worth pointing out some new activity on the issue.

**Comment by jeremiah-corrado** [comment]
I'd like to express interest in the `movedTypeNeedsDestroy` and `destroyMoved` names that Ben suggested above.

I think these have a couple of advantages:
- the name of the module itself, and the other procedures also have the word "move" in them. This symmetry is nice, and helps to convey the relationship between all the procedures.
- if  saw these procedures in some code, and was unfamiliar with them, my first question would be "what does 'moved' mean in this context?". And I think this is the correct question to be asking when approaching the `MemMove` docs.
    - in other words, it might indicate to them that these are to be used in a very specialized context (when the author of a container is taking manual control over memory allocation) and not in typical Chapel code.
    - alternatively, reading some of the other names might make me question my understanding of how memory management works in Chapel. I.e., "do I always need to check if a value requires manual destruction when I'm done with it?"

That being said, it sounds like those names may have already been ruled out during an offline discussion. I'd be curious to hear what the downsides are.

**Comment by dlongnecke-cray** [comment]
I think `movedType` is kind of a misnomer because I can write something like:

```chapel
var x: someRecord;
if movedTypeNeedsDestroy(x.type) then destroyMoved(x);
```

But I didn't really _move_ `x` at all. It's just sitting there. I also didn't move its type (how can you move that which has no runtime value?!?!).

I'm also not a fan of `destroy` because that terminology reminds me of C++, and I thought we were trying to separate ourselves from evoking images of C++ destructors. At least, I've tried to consistent about using "deinitialize" as the term owing to that line of reasoning.

I'm not sure that we need to reintroduce that term to abstract away, here. I think `deinitialize` works because what we really want is to avoid suggesting that you are calling _deinit_ (and deinitialization does not necessarily involve calling a `deinit` function - e.g., Michael's example about managed classes - they're implemented with records today, but the compiler could handle their deinitialization more directly in another implementation).

> the name of the module itself, and the other procedures also have the word "move" in them. 

I agree this symmetry is nice, but since Chapel has first-class namespacing through modules, I expect you'd do something like this instead to preserve symmetry:

```chapel
import MemMove;
if MemMove.typeNeedsDestroy(t) then ...;
```

> alternatively, reading some of the other names might make a new user question their understanding of how memory management works in Chapel. 

Yeah I think this is unavoidable to a degree, for such a low-level function. I think the best we can do is avoid the implication that `explicitDeinit`/`destroyMoved`/`deinitialize` is actually calling a `deinit` function, as that may not be the case. I think either way the docs are still going to have to explain the consequences of deinitializing different categories of type (e.g., `int` vs `record` vs `class`).

**Comment by jeremiah-corrado** [comment]
> I think movedType is kind of a misnomer because I can write something like:
> ```chapel
> var x: someRecord;
> if movedTypeNeedsDestroy(x.type) then destroyMoved(x);
> ```
> But I didn't really move x at all. It's just sitting there.

Ok, that makes sense. Out of curiosity, is the above code something that you would ever need to write in practice? I thought that this functionality was typically only used if a value had been moved into some memory location that was malloced manually, like a `_ddata` array?

**Comment by mppf** [comment]
FWIW one might need to use these functions with a `malloc`'d buffer, even though they are not moving elements in to / out of the buffer. (The exact pattern with a `var` seems unlikely, but it could come up if the author of the code is trying to use low-level features to optimize copies).

**Comment by dlongnecke-cray** [comment]
Yeah the example is contrived, but what I was trying to get at was that it doesn't really seem like this deinitialization function is related to `move` operations other than tangentially. You're right that the common pattern will be deinitializing a value that was moved into a buffer.

**Comment by benharsh** [comment]
The subteam for this topic met today and resolved the open questions regarding this issue. In summary:

1. We have reached consensus for the following names: ``needsDestroy`` and ``destroy``. (More on this below)
```chpl
proc needsDestroy(type t) param : bool
proc destroy(ref obj: ?t)
```

2. We have elected not to introduce a compiler error for the ``destroy`` method, as it seems of minimal or zero utility when it comes to user productivity and clarifying some misunderstanding by the user.

3. ``destroy`` will be a no-op for unmanaged classes.

---

On the topic of naming, there was a general concern about the potential confusion surrounding the term ``deinit``, and whether that might lead users to believe that ``unmanaged`` or ``shared`` classes would be fully destroyed (not true for unmanaged, possible for shared). There was some discussion of using the term "deinitialization" to distinguish the ``deinit`` method from the process of destroying an object, but opinions varied as to whether this difference was sufficient or if it was just as confusing as ``deinit``.

Concern over the term "destroy" was discussed, in which it was pointed out that the notion of "destroy" might invoke thoughts of C++ destructors, which we have gone through some effort to distance ourselves from after having deprecated our old destructor implementation in the past. There was also concern over the potentially different terms we might use in our documentation to describe the way in which variable are "cleaned up" by the compiler. We tend to use "deinitialize" or "deinitialization" in various places in the language specification, and also use "destroy" in relatively fewer places.

We also discussed borrowing some terms or naming-styles from other languages like Rust, but could not land on an option with much support. For example, rust supports ``needsDrop`` and ``dropInPlace``, which might have worked well with ``deinit`` replacing ``drop``, if not for our general hesitance towards ``deinit``. The term ``reclaim`` was also proposed, but did not gain support over ``destroy``.

In moving forward, we concluded that we either supported or could live with the "destroy" option. We will need to re-examine our wording in documentation and the language specification, which may help to mitigate concern for too many terms describing the same thing. We will also need to clearly document these two functions and the situations or types where they are applicable.

Thanks to @jeremiah-corrado , @mppf , @dlongnecke-cray for their input!
